core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -62,12 +62,54 @@ object AdminUtils extends Logging {
    * p8        p9        p5        p6        p7       (2nd replica)
    * p3        p4        p0        p1        p2       (3nd replica)
    * p7        p8        p9        p5        p6       (3nd replica)
+   *
+   * To create rack aware assignment, this API will first create an rack interlaced broker list. For example,
+   * from this brokerID -> rack mapping:
+   *
+   * 0 -> "rack1", 1 -> "rack3", 2 -> "rack3", 3 -> "rack2", 4 -> "rack2", 5 -> "rack1"
+   *
+   * The rack interlaced list will be
+   *
+   * 0, 3, 1, 5, 4, 2
+   *
+   * Then an easy round-robin assignment can be applied. Assume 6 partitions with replication factor of 3, the assignment
+   * will be
+   *
+   * 0 -> 0,3,1
+   * 1 -> 3,1,5
+   * 2 -> 1,5,4
+   * 3 -> 5,4,2
+   * 4 -> 4,2,0
+   * 5 -> 2,0,3
+   *
+   * Once it has completed the first round-robin, if there are more partitions to assign, the algorithm will start to
+   * have a shift for the followers. This is to ensure we will not always get the same set of sequence.
+   * In this case, if there is another partition to assign (partition #6), the assignment will be
+   *
+   * 6 -> 0,4,2 (instead of repeating 0,3,1 as partition 0)
+   *
+   * The rack aware assignment always chooses leader of the partition using round robin on the rack interlaced broker list.
+   * When choosing followers, it will be biased towards brokers on racks that do not have
+   * any replica assignment, until every rack has a replica. Then the follower assignment will go back to round-robin on
+   * the broker list.
+   *
+   * As the result, if the number of replicas is equal to or greater than the number of racks, it will ensure that
+   * each rack will get at least one replica. Otherwise, each rack will get
+   * at most one replica. In the perfect situation where number of replica is the same as number of racks and each rack
+   * has the same number of brokers, it guarantees that replica distribution is even across brokers and racks.
+   *
+   * @param rackInfo Map from broker ID to its rack (zone). If empty, no rack aware
+   *                 assignment will be used.
+   * @throws AdminOperationException If rack information is supplied but is incomplete, of if
+   *                                 there is not possible to assign each replica to a unique rack
+   *
    */
   def assignReplicasToBrokers(brokerList: Seq[Int],
                               nPartitions: Int,
                               replicationFactor: Int,
                               fixedStartIndex: Int = -1,
-                              startPartitionId: Int = -1)
+                              startPartitionId: Int = -1,
+                              rackInfo: Map[Int, String] = Map())
```

Comment:
rackInfo here can be different from the actual broker-rack mapping. In case some brokers have rack and some brokers do not have rack, AdminUtils.getBrokersAndRackInfo() will modify the mapping depending on how strict we want to be rack aware. It will also return empty Map if user does not want rack aware. I think it is better to have higher level API (like createTopic()) to be influenced by the rack aware mode depending on the situation and user input and leave this assignment API free of that influence.